OpenMage/magento-lts
# Fix PHP warnings when session backend exits before returning.

### Description

When a session handler exits during `session_start()` the PHP engine still tries to end the session during shutdown even though it was never successfully started. The `Cm_RedisSession` was updated with a `setDieOnError` method (requires 3.1+) to allow this automatic exit to be disabled and this patch for OpenMage disables it and then catches any thrown exceptions and handles the exit more gracefully by first calling `session_abort()`.

### Fixed Issues

PHP throws a few warnings when exiting un-gracefully during `session_start()`:

- Warning: Unknown: Failed to write session data using user defined save handler. (session.save_pat...
- Warning: Unknown: Cannot call session save handler in a recursive manner  in Unknown on line 0

### Manual testing scenarios

Setup Redis sessions in `app/etc/local/xml`:

```
<config>
    ...
    <global>
        ...
        <redis_session>
            <host>redis</host>
            <port>6379</port>
            <max_concurrency>3</max_concurrency>
        </redis_session>
    </global>
</config>
```

Pass a number of concurrent requests to overwhelm the concurrency limit:

```
hey -c 10 -n 10 -H 'Content-Type: application/x-www-form-urlencoded' \
  -H 'Cookie: XDEBUG_SESSION=phpstorm; adminhtml=647625dfe137cf209889b524fb177e6b' \
  -H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36' \
  'http://openmage-7f000001.nip.io/index.php/admin/index/index/key/955d7347e9813e7994ca43c8b054003a/'
```
Summary:
```
Status code distribution:
  [200] 3 responses
  [503] 7 responses
```

Observe that a 503 error is returned for the excess requests and check if any warnings are logged due to the session backend errors.
